File: sellar/lib/src/features/products/data/product_repository.dart
import 'package:sellar/src/constants/app_constants.dart';
import 'package:sellar/src/features/products/domain/entities/product.dart';
import 'package:sellar/src/services/api_service.dart';

/// Repository for product CRUD operations
class ProductRepository {
  ProductRepository({required ApiService apiService})
      : _apiService = apiService;

  final ApiService _apiService;

  Future<List<Product>> getProducts() async {
    final response = await _apiService.get(AppConstants.productsPath);
    if (response.statusCode == 200 && response.data != null) {
      final list = response.data['data'] as List<dynamic>;
      return list
          .map((e) => Product.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    throw Exception(response.data?['error'] ?? 'Failed to fetch products');
  }

  Future<Product> createProduct({
    required String name,
    required String description,
    required String category,
    required double price,
    int? quantity,
    List<String> tags = const [],
    List<String> images = const [],
  }) async {
    final response = await _apiService.post(
      AppConstants.productsPath,
      data: {
        'name': name,
        'description': description,
        'category': category,
        'price': price,
        if (quantity != null) 'quantity': quantity,
        'tags': tags,
        'images': images,
      },
    );
    if (response.statusCode == 201 && response.data != null) {
      return Product.fromJson(response.data['data'] as Map<String, dynamic>);
    }
    throw Exception(response.data?['error'] ?? 'Failed to create product');
  }

  Future<Product> updateProduct(
    String id, {
    String? name,
    String? description,
    String? category,
    double? price,
    int? quantity,
    List<String>? tags,
    List<String>? images,
    bool? isActive,
  }) async {
    final response = await _apiService.put(
      '${AppConstants.productsPath}/$id',
      data: {
        if (name != null) 'name': name,
        if (description != null) 'description': description,
        if (category != null) 'category': category,
        if (price != null) 'price': price,
        if (quantity != null) 'quantity': quantity,
        if (tags != null) 'tags': tags,
        if (images != null) 'images': images,
        if (isActive != null) 'isActive': isActive,
      },
    );
    if (response.statusCode == 200 && response.data != null) {
      return Product.fromJson(response.data['data'] as Map<String, dynamic>);
    }
    throw Exception(response.data?['error'] ?? 'Failed to update product');
  }

  Future<void> deleteProduct(String id) async {
    final response =
        await _apiService.delete('${AppConstants.productsPath}/$id');
    if (response.statusCode != 200) {
      throw Exception(response.data?['error'] ?? 'Failed to delete product');
    }
  }
}

File: sellar-api/src/repositories/productRepository.ts
import { prisma } from '../utils/database';

export class ProductRepository {
  async findAllByVendor(vendorId: string) {
    return prisma.product.findMany({
      where: { vendorId },
      orderBy: { createdAt: 'desc' },
    });
  }

  async findOneByVendor(id: string, vendorId: string) {
    return prisma.product.findFirst({ where: { id, vendorId } });
  }

  async create(vendorId: string, data: {
    name: string;
    description: string;
    category: string;
    tags: string[];
    price: number;
    quantity?: number | null;
    images: string[];
  }) {
    return prisma.product.create({ data: { vendorId, ...data } });
  }

  async update(id: string, data: {
    name?: string;
    description?: string;
    category?: string;
    tags?: string[];
    price?: number;
    quantity?: number | null;
    images?: string[];
    isActive?: boolean;
  }) {
    return prisma.product.update({ where: { id }, data });
  }

  async delete(id: string) {
    return prisma.product.delete({ where: { id } });
  }
}
